File: engines/ts/src/bundle-signature.ts
/**
 * Optional Ed25519 detached signatures for harness bundles (M10.4).
 *
 * Covered bytes: UTF-8 of canonicalJson(bundle without `signature`).
 * Signature object: { alg: "ed25519", public_key: hex32, sig: hex64 }.
 */
import {
  createPrivateKey,
  createPublicKey,
  generateKeyPairSync,
  sign as cryptoSign,
  verify as cryptoVerify,
  type KeyObject,
} from "node:crypto";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { canonicalJson } from "./canonical.js";

export const BUNDLE_SIG_ALG = "ed25519" as const;

/** PKCS8 DER prefix for a 32-byte Ed25519 seed. */
const ED25519_PKCS8_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
/** SPKI DER prefix for a 32-byte Ed25519 public key. */
const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export type BundleSignature = {
  alg: typeof BUNDLE_SIG_ALG;
  public_key: string;
  sig: string;
};

export class BundleSignatureError extends Error {
  readonly code:
    | "bundle_signature_invalid"
    | "bundle_signature_unsupported"
    | "bundle_sign_key_invalid";
  constructor(code: BundleSignatureError["code"], message: string) {
    super(message);
    this.name = "BundleSignatureError";
    this.code = code;
  }
}

const HEX32_RE = /^[a-f0-9]{64}$/;
const HEX64_RE = /^[a-f0-9]{128}$/;

function privateKeyFromSeed(seed: Buffer): KeyObject {
  if (seed.length !== 32) {
    throw new BundleSignatureError(
      "bundle_sign_key_invalid",
      `Ed25519 seed must be 32 bytes, got ${seed.length}`,
    );
  }
  return createPrivateKey({
    key: Buffer.concat([ED25519_PKCS8_PREFIX, seed]),
    format: "der",
    type: "pkcs8",
  });
}

function publicKeyFromRaw(raw: Buffer): KeyObject {
  if (raw.length !== 32) {
    throw new BundleSignatureError(
      "bundle_signature_invalid",
      `Ed25519 public key must be 32 bytes, got ${raw.length}`,
    );
  }
  return createPublicKey({
    key: Buffer.concat([ED25519_SPKI_PREFIX, raw]),
    format: "der",
    type: "spki",
  });
}

function rawPublicKey(privateKey: KeyObject): Buffer {
  const pub = createPublicKey(privateKey);
  return pub.export({ type: "spki", format: "der" }).subarray(12);
}

/** Strip signature and return the object that is signed. */
export function signingPayload(bundle: Record<string, unknown>): Record<string, unknown> {
  const { signature: _sig, ...rest } = bundle;
  return rest;
}

/** UTF-8 bytes of canonical JSON of the unsigned payload. */
export function signingBytes(bundle: Record<string, unknown>): Buffer {
  return Buffer.from(canonicalJson(signingPayload(bundle)), "utf8");
}

export function parseBundleSignature(raw: unknown): BundleSignature {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new BundleSignatureError("bundle_signature_invalid", "bundle.signature must be an object");
  }
  const obj = raw as Record<string, unknown>;
  if (obj.alg !== BUNDLE_SIG_ALG) {
    throw new BundleSignatureError(
      "bundle_signature_unsupported",
      `unsupported signature alg: ${String(obj.alg)}`,
    );
  }
  if (typeof obj.public_key !== "string" || !HEX32_RE.test(obj.public_key)) {
    throw new BundleSignatureError(
      "bundle_signature_invalid",
      "bundle.signature.public_key must be 64-char lowercase hex (32-byte Ed25519 key)",
    );
  }
  if (typeof obj.sig !== "string" || !HEX64_RE.test(obj.sig)) {
    throw new BundleSignatureError(
      "bundle_signature_invalid",
      "bundle.signature.sig must be 128-char lowercase hex (64-byte Ed25519 signature)",
    );
  }
  const allowed = new Set(["alg", "public_key", "sig"]);
  for (const k of Object.keys(obj)) {
    if (!allowed.has(k)) {
      throw new BundleSignatureError(
        "bundle_signature_invalid",
        `unexpected signature field: ${k}`,
      );
    }
  }
  return {
    alg: BUNDLE_SIG_ALG,
    public_key: obj.public_key,
    sig: obj.sig,
  };
}

/** Verify detached signature when present. No-op when signature is absent. */
export function verifyBundleSignature(bundle: Record<string, unknown>): void {
  if (!("signature" in bundle) || bundle.signature === undefined) return;
  const signature = parseBundleSignature(bundle.signature);
  const ok = cryptoVerify(
    null,
    signingBytes(bundle),
    publicKeyFromRaw(Buffer.from(signature.public_key, "hex")),
    Buffer.from(signature.sig, "hex"),
  );
  if (!ok) {
    throw new BundleSignatureError(
      "bundle_signature_invalid",
      "bundle signature verification failed",
    );
  }
}

export function loadSignKeyMaterial(input: string | Buffer): KeyObject {
  let raw: Buffer;
  if (typeof input === "string") {
    const trimmed = input.trim();
    if (trimmed.includes("BEGIN") && trimmed.includes("PRIVATE KEY")) {
      return createPrivateKey(trimmed);
    }
    if (HEX32_RE.test(trimmed.toLowerCase().replace(/\s+/g, ""))) {
      raw = Buffer.from(trimmed.toLowerCase().replace(/\s+/g, ""), "hex");
      return privateKeyFromSeed(raw);
    }
    // Treat as filesystem path
    const abs = resolve(trimmed);
    const file = readFileSync(abs);
    return loadSignKeyMaterial(file);
  }
  const text = input.toString("utf8").trim();
  if (text.includes("BEGIN") && text.includes("PRIVATE KEY")) {
    return createPrivateKey(text);
  }
  const hex = text.toLowerCase().replace(/\s+/g, "");
  if (HEX32_RE.test(hex)) {
    return privateKeyFromSeed(Buffer.from(hex, "hex"));
  }
  if (input.length === 32) {
    return privateKeyFromSeed(input);
  }
  // Try DER PKCS8
  try {
    return createPrivateKey({ key: input, format: "der", type: "pkcs8" });
  } catch {
    throw new BundleSignatureError(
      "bundle_sign_key_invalid",
      "sign key must be 32-byte seed (raw or 64-hex), or PKCS8 PEM/DER Ed25519 private key",
    );
  }
}

export function loadSignKeyFromPath(path: string): KeyObject {
  return loadSignKeyMaterial(path);
}

/** Attach Ed25519 signature to a bundle object (mutates a shallow copy). */
export function signBundle(
  bundle: Record<string, unknown>,
  privateKey: KeyObject | string | Buffer,
): Record<string, unknown> {
  const key =
    typeof privateKey === "string" || Buffer.isBuffer(privateKey)
      ? loadSignKeyMaterial(privateKey)
      : privateKey;
  const unsigned = signingPayload(bundle);
  const sig = cryptoSign(null, Buffer.from(canonicalJson(unsigned), "utf8"), key);
  const public_key = rawPublicKey(key).toString("hex");
  return {
    ...unsigned,
    signature: {
      alg: BUNDLE_SIG_ALG,
      public_key,
      sig: sig.toString("hex"),
    } satisfies BundleSignature,
  };
}

/** Generate a new Ed25519 seed (hex) + matching public key hex for tests/docs. */
export function generateEd25519SeedHex(): { seed_hex: string; public_key: string } {
  const { privateKey } = generateKeyPairSync("ed25519");
  const der = privateKey.export({ type: "pkcs8", format: "der" });
  const seed = der.subarray(der.length - 32);
  return {
    seed_hex: seed.toString("hex"),
    public_key: rawPublicKey(privateKey).toString("hex"),
  };
}

File: engines/py/fausth/bundle_signature.py
"""Optional Ed25519 detached signatures for harness bundles (M10.4).

Covered bytes: UTF-8 of canonical_json(bundle without ``signature``).
Signature object: ``{alg: "ed25519", public_key: hex32, sig: hex64}``.
"""
from __future__ import annotations

import re
from pathlib import Path
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .canonical import canonical_json

BUNDLE_SIG_ALG = "ed25519"

HEX32_RE = re.compile(r"^[a-f0-9]{64}$")
HEX64_RE = re.compile(r"^[a-f0-9]{128}$")


class BundleSignatureError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


def signing_payload(bundle: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in bundle.items() if k != "signature"}


def signing_bytes(bundle: dict[str, Any]) -> bytes:
    return canonical_json(signing_payload(bundle)).encode("utf-8")


def parse_bundle_signature(raw: Any) -> dict[str, str]:
    if not isinstance(raw, dict):
        raise BundleSignatureError("bundle_signature_invalid", "bundle.signature must be an object")
    alg = raw.get("alg")
    if alg != BUNDLE_SIG_ALG:
        raise BundleSignatureError(
            "bundle_signature_unsupported",
            f"unsupported signature alg: {alg!r}",
        )
    public_key = raw.get("public_key")
    if not isinstance(public_key, str) or not HEX32_RE.fullmatch(public_key):
        raise BundleSignatureError(
            "bundle_signature_invalid",
            "bundle.signature.public_key must be 64-char lowercase hex (32-byte Ed25519 key)",
        )
    sig = raw.get("sig")
    if not isinstance(sig, str) or not HEX64_RE.fullmatch(sig):
        raise BundleSignatureError(
            "bundle_signature_invalid",
            "bundle.signature.sig must be 128-char lowercase hex (64-byte Ed25519 signature)",
        )
    allowed = {"alg", "public_key", "sig"}
    for k in raw:
        if k not in allowed:
            raise BundleSignatureError(
                "bundle_signature_invalid",
                f"unexpected signature field: {k}",
            )
    return {"alg": BUNDLE_SIG_ALG, "public_key": public_key, "sig": sig}


def verify_bundle_signature(bundle: dict[str, Any]) -> None:
    if "signature" not in bundle or bundle["signature"] is None:
        return
    signature = parse_bundle_signature(bundle["signature"])
    pub = Ed25519PublicKey.from_public_bytes(bytes.fromhex(signature["public_key"]))
    try:
        pub.verify(bytes.fromhex(signature["sig"]), signing_bytes(bundle))
    except InvalidSignature as e:
        raise BundleSignatureError(
            "bundle_signature_invalid",
            "bundle signature verification failed",
        ) from e


def _private_key_from_seed(seed: bytes) -> Ed25519PrivateKey:
    if len(seed) != 32:
        raise BundleSignatureError(
            "bundle_sign_key_invalid",
            f"Ed25519 seed must be 32 bytes, got {len(seed)}",
        )
    return Ed25519PrivateKey.from_private_bytes(seed)


def load_sign_key_material(input: str | bytes | Path) -> Ed25519PrivateKey:
    if isinstance(input, Path):
        return load_sign_key_material(input.read_bytes())
    if isinstance(input, str):
        trimmed = input.strip()
        if "BEGIN" in trimmed and "PRIVATE KEY" in trimmed:
            key = serialization.load_pem_private_key(trimmed.encode("utf-8"), password=None)
            if not isinstance(key, Ed25519PrivateKey):
                raise BundleSignatureError(
                    "bundle_sign_key_invalid",
                    "PEM key must be an Ed25519 private key",
                )
            return key
        hex_body = re.sub(r"\s+", "", trimmed).lower()
        if HEX32_RE.fullmatch(hex_body):
            return _private_key_from_seed(bytes.fromhex(hex_body))
        path = Path(trimmed)
        if path.is_file():
            return load_sign_key_material(path.read_bytes())
        raise BundleSignatureError(
            "bundle_sign_key_invalid",
            f"sign key path not found or invalid material: {trimmed}",
        )

    text = input.decode("utf-8", errors="ignore").strip()
    if "BEGIN" in text and "PRIVATE KEY" in text:
        key = serialization.load_pem_private_key(text.encode("utf-8"), password=None)
        if not isinstance(key, Ed25519PrivateKey):
            raise BundleSignatureError(
                "bundle_sign_key_invalid",
                "PEM key must be an Ed25519 private key",
            )
        return key
    hex_body = re.sub(r"\s+", "", text).lower()
    if HEX32_RE.fullmatch(hex_body):
        return _private_key_from_seed(bytes.fromhex(hex_body))
    if len(input) == 32:
        return _private_key_from_seed(input)
    try:
        key = serialization.load_der_private_key(input, password=None)
        if not isinstance(key, Ed25519PrivateKey):
            raise BundleSignatureError(
                "bundle_sign_key_invalid",
                "DER key must be an Ed25519 private key",
            )
        return key
    except BundleSignatureError:
        raise
    except Exception as e:
        raise BundleSignatureError(
            "bundle_sign_key_invalid",
            "sign key must be 32-byte seed (raw or 64-hex), or PKCS8 PEM/DER Ed25519 private key",
        ) from e


def load_sign_key_from_path(path: str | Path) -> Ed25519PrivateKey:
    return load_sign_key_material(Path(path))


def _raw_public_hex(private_key: Ed25519PrivateKey) -> str:
    return (
        private_key.public_key()
        .public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )
        .hex()
    )


def sign_bundle(
    bundle: dict[str, Any],
    private_key: Ed25519PrivateKey | str | bytes | Path,
) -> dict[str, Any]:
    if not isinstance(private_key, Ed25519PrivateKey):
        key = load_sign_key_material(private_key)
    else:
        key = private_key
    unsigned = signing_payload(bundle)
    sig = key.sign(canonical_json(unsigned).encode("utf-8"))
    return {
        **unsigned,
        "signature": {
            "alg": BUNDLE_SIG_ALG,
            "public_key": _raw_public_hex(key),
            "sig": sig.hex(),
        },
    }


def generate_ed25519_seed_hex() -> dict[str, str]:
    key = Ed25519PrivateKey.generate()
    seed = key.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    return {"seed_hex": seed.hex(), "public_key": _raw_public_hex(key)}
